library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/RecipeLoadingEvents.java

```diff
@@ -0,0 +1,285 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.recipe.api;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.recipe.Recipe;
+import net.minecraft.recipe.RecipeType;
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.base.api.event.EventAwareListener;
+
+/**
+ * Represents the recipe loading events.
+ * <p>
+ * Triggered when the recipes are loaded.
+ *
+ * @version 1.0.0
+ * @since 1.0.0
+ */
+public final class RecipeLoadingEvents {
+	/**
+	 * Recipe loading event, triggered when the recipes are loaded.
+	 */
+	public static final Event<RecipeLoadingCallback> REGISTER = Event.create(RecipeLoadingCallback.class,
+			callbacks -> handler -> {
+				for (var callback : callbacks) {
+					callback.onLoadRecipe(handler);
+				}
+			});
+	public static final Event<RecipeModifyCallback> MODIFY = Event.create(RecipeModifyCallback.class,
+			callbacks -> handler -> {
+				for (var callback : callbacks) {
+					callback.onModifyRecipe(handler);
+				}
+			});
+	public static final Event<RecipeRemoveCallback> REMOVE = Event.create(RecipeRemoveCallback.class,
+			callbacks -> handler -> {
+				for (var callback : callbacks) {
+					callback.onRecipeRemovePhase(handler);
+				}
+			});
+
+	private RecipeLoadingEvents() {
+		throw new UnsupportedOperationException("RecipeLoadingEvents only contains static definitions.");
+	}
+
+	/**
+	 * Callback called to register additional recipes when recipes are loaded.
+	 */
+	@FunctionalInterface
+	public interface RecipeLoadingCallback extends EventAwareListener {
+		/**
+		 * Called when recipes are loaded.
+		 * <p>
+		 * {@code handler} is used to add recipes into the {@linkplain net.minecraft.recipe.RecipeManager recipe manager}.
+		 *
+		 * @param handler the recipe handler
+		 */
+		void onLoadRecipe(RecipeHandler handler);
+
+		/**
+		 * This interface should not be extended by users.
+		 */
+		@ApiStatus.NonExtendable
+		interface RecipeHandler {
+			/**
+			 * Registers a recipe into the {@link net.minecraft.recipe.RecipeManager}.
+			 * <p>
+			 * The recipe factory is only called if the recipe can be registered.
+			 *
+			 * @param id      identifier of the recipe
+			 * @param factory the recipe factory
+			 */
+			void register(Identifier id, Function<Identifier, Recipe<?>> factory);
+		}
+	}
+
+	/**
+	 * Callback called to modify or replace recipes after recipes are loaded.
+	 */
+	@FunctionalInterface
+	public interface RecipeModifyCallback extends EventAwareListener {
+		/**
+		 * Called after recipes are loaded to modify and replace recipes.
+		 *
+		 * @param handler the recipe handler
+		 */
+		void onModifyRecipe(RecipeHandler handler);
```

Comment:
same as above, this seems to be an API for modifying recipes, but the method name implies this is used to listen for when a mod modifies a recipe.